agent/hcp/client/telemetry_config.go

```diff
@@ -0,0 +1,170 @@
+	}
+
+	telemetryConfig := resp.Payload.TelemetryConfig
+
+	metricsEndpoint, err := convertMetricEndpoint(telemetryConfig.Endpoint, telemetryConfig.Metrics.Endpoint)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse metrics endpoint: %w", err)
+	}
+
+	metricsFilters, err := convertMetricFilters(telemetryConfig.Metrics.IncludeList)
+	if err != nil {
+		// Do not fail on bad regex, as we can update these later on dynamically.
+		logger.Error("failed to parse regex filters", "error", err)
+	}
+
+	metricLabels := convertMetricLabels(telemetryConfig.Labels, cfg)
+
+	return &TelemetryConfig{
+		MetricsConfig: &MetricsConfig{
+			Endpoint: metricsEndpoint,
+			Labels:   metricLabels,
+			Filters:  metricsFilters,
+		},
+		RefreshConfig: &RefreshConfig{
+			RefreshInterval: refreshInterval,
+		},
+	}, nil
+}
+
+// convertMetricEndpoint returns a url for the export of metrics, if a valid endpoint was obtained.
+// It returns no error, and no url, if an empty endpoint is retrieved (server not registered with CCM).
+// It returns an error, and no url, if a bad endpoint is retrieved.
+func convertMetricEndpoint(telemetryEndpoint string, metricsEndpoint string) (*url.URL, error) {
+	// Telemetry endpoint overriden by metrics specific endpoint, if given.
+	endpoint := telemetryEndpoint
+	if metricsEndpoint != "" {
+		endpoint = metricsEndpoint
+	}
+
+	// If endpoint is empty, server not registered with CCM, no error returned.
+	if endpoint == "" {
+		return nil, nil
+	}
+
+	// Endpoint from CTW has no metrics path, so it must be added.
+	rawUrl := endpoint + metricsGatewayPath
+	u, err := url.ParseRequestURI(rawUrl)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse url: %w", err)
+	}
+
+	return u, nil
+}
+
+// convertMetricFilters returns a valid regex used to filter metrics.
+// It fails if there are 0 valid regex filters given.
+func convertMetricFilters(payloadFilters []string) (*regexp.Regexp, error) {
+	var mErr error
+	filters := payloadFilters
+	validFilters := make([]string, 0, len(filters))
+	for _, filter := range filters {
+		_, err := regexp.Compile(filter)
+		if err != nil {
+			mErr = multierror.Append(mErr, fmt.Errorf("compilation of filter %q failed: %w", filter, err))
+			continue
+		}
+		validFilters = append(validFilters, filter)
+	}
+
+	if len(validFilters) == 0 {
+		return nil, multierror.Append(mErr, fmt.Errorf("no valid filters"))
+	}
+
+	// Combine the valid regex strings with OR.
+	finalRegex := strings.Join(validFilters, "|")
+	composedRegex, err := regexp.Compile(finalRegex)
+	if err != nil {
+		return nil, fmt.Errorf("failed to compile regex: %w", err)
+	}
+
+	return composedRegex, nil
+}
+
+// convertMetricLabels returns a set of <key, value> string pairs that must be added as attributes to all exported telemetry data.
+func convertMetricLabels(payloadLabels map[string]string, cfg config.CloudConfig) map[string]string {
+	labels := make(map[string]string)
```

Comment:
Yeah I remembered us discussing this, one reason why I think the current solution was chosen is because there is a potential for payloadLabels to be `nil`? i